### ISSUE TITLE
Suggestion how to separate Views for Page related types.

### ISSUE THREAD

**Issue by andorp** [liqd/aula#20]
The `Frontend.Html` should contain some view related functionality.
````haskell
-- | Page View for Digestive Functors
class PageView a where
    pageView :: View Html -> ST -> a -> Html

instance PageView PageHomeWithLoginPrompt where
    pageView v formAction PageHomeWithLoginPrompt =
        DF.form v formAction $ do
            DF.div $
                DF.h1 "Willkommen bei Aula"
            DF.div $ do
                DF.inputText "login" v
                DF.inputText "pass" v
                DF.inputSubmit "ANMELDEN"
            DF.div $
                DF.p "Solltest du dein Password nicht mehr kennen, melde dich bitte bei den Admins eurer Schule."
````
The `Frontend` module could import and use the views.
````haskell
loginForm :: Server (FormH HTML Html ())
loginForm = formH "/login" validate process render
  where
    render :: View Html -> ST -> ExceptT ServantErr IO Html
    render v formAction = pure $ pageView v PageHomeWithLoginPrompt

    validate :: DF.Form Html (ExceptT ServantErr IO) (ST, ST)
    validate = (,)
        <$> ("login" .: DF.text (Just "Bein benutzername"))
        <*> ("pass"  .: DF.text (Just "Dein Password"))

    process :: (ST, ST) -> ExceptT ServantErr IO Html
    process _ = do
        return . toMarkup . Frame . text $ "home"
````

Please share your thoughts.

Edited by @andorp

**Comment by Mikolaj** [comment]
I'm only starting to grok the magic of our tech stack, but the proposal looks very promising. Nitpicks: there is a missing ``v`` argument in ``render`` and you also failed to pass ``formAction`` through, which is currently unbound in ``pageView`` (do we need another parameter there?).

**Comment by andorp** [comment]
@Mikolaj You are right, I missed to pass the argument v, I had just sketched the proposal.

**Comment by andorp** [comment]
@Mikolaj You are right about the form action as well. I am not sure if the form action would come from the caller side of the `pageView`. Probably it is related to the paths we give in the servant path description.

**Comment by fisx** [comment]
> Probably it is related to the paths we give in the servant path description.

you mean in the type-level routing table?  yes, it's actually the same in most cases.

servant offers safe links somewhere, but i looked at them and wasn't satisfied they solve this awkwardness.  I'm also not too worried about it: we need at least one test per form anyway that checks that the post action has some specific response and effect, and that should detect all inconsistencies between routing table and action strings.

**Comment by andorp** [comment]
> you mean in the type-level routing table? 
Yes. Let's think about how we can put as much as check we can to the type level as we are lack of QA.

**Comment by andorp** [comment]
I started to experiment with the `class PageView a` . I would like to see how deep is the rabbit's hole.

**Comment by fisx** [comment]
closed by #41
